client/src/main/java/io/pravega/client/tables/TableTransaction.java

```diff
@@ -0,0 +1,83 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.tables;
+
+import com.google.common.annotations.Beta;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Defines all operations that can be made on a Table Transaction.
+ *
+ * Notes about Updates:
+ * * Updates (Conditional or not) in this TableTransaction will not be reflected in the main Table until {@link #commit()}
+ * is invoked (and completes successfully).
+ * * Conditional Updates in this TableTransaction can only be done by comparing with Keys in this Transaction. Comparisons
+ * cannot be made against Keys from the main Table.
+ * * {@link #commit()} cannot be made conditional against the state of the main Table.
+ *
+ * @param <KeyT>   Table Key Type.
+ * @param <ValueT> Table Value Type.
+ */
+@Beta
+public interface TableTransaction<KeyT, ValueT> extends AutoCloseable {
+    /**
+     * Performs an Unconditional Update by inserting or updating the Value for a Key in this Table, regardless of whether
+     * the Key exists or not or what its Version is. See {@link TableWriter#put(KeyT, ValueT)}.
+     *
+     * @param key   The Key to insert or update.
+     * @param value The Value to set for the Key.
+     * @return A CompletableFuture that, when completed, will contain the {@link KeyVersion} associated with the newly
+     * inserted or updated entry.
+     */
+    CompletableFuture<KeyVersion> put(KeyT key, ValueT value);
+
+    /**
+     * Performs a Conditional Update by inserting or updating the Value for a Key in this Table.
+     * See {@link TableWriter#put(KeyT, ValueT, KeyVersion)}.
+     *
+     * @param key            The Key to insert or update.
+     * @param value          The Value to set for the Key.
+     * @param compareVersion A {@link KeyVersion} that will need to match the version of the Key in the Table. Consider
+     *                       using {@link KeyVersion#NOT_EXISTS} if this is used to insert a Key which must not previously
+     *                       exist.
+     * @return A CompletableFuture that, when completed, will contain the {@link KeyVersion} associated with the newly
+     * inserted or updated entry.
+     */
+    CompletableFuture<KeyVersion> put(KeyT key, ValueT value, KeyVersion compareVersion);
```

Comment:
As stated above, the condition is for the `put` operation, not for the merge. Maybe someone wants to do conditional updates inside the transaction to ensure the transaction itself is in a good shape.

Reply:
It's a weird position to be in, because you didn't actually supply a get() api on the transaction. (For understandable reasons) So if there are multiple processes attempting to operate on the same transactions, it doesn't really work. And if it's just a single process, its still odd because there is no way to compare against the value having not been changed in the txn or against a particular value existing in the main segment at the time of creation. So it's just a nitch usecase where for some reason a process knows it wrote something but is unsure what it wrote to the transaction. I don't want to clutter up our public API for such a rare case.